test/tests/functional/pbs_array_job.py

```diff
@@ -0,0 +1,72 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class TestArrayJobs(TestFunctional):
+
+    """
+    This test suite tests if PBS's array job functionality
+    """
+    def test_expired_subjobs_not_reported(self):
+        """
+        Test if a subjob is finished ans moves to expired state,
+        it is not reported to scheduler in the next scheduling
+        cycle. Scheduler expects only running subjobs to be reported to it.
+        """
+
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        a = {'job_history_enable': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        res_req = {'Resource_List.select': '1:ncpus=1',
+                   'array_indices_submitted': '1-6'}
+        j1 = Job(TEST_USER, attrs=res_req)
+        j1.set_sleep_time(2)
```

Comment:
2 seconds seems to small to me for slower machines, how about turning scheduling off after submitting the job, doing expect to see if it expired, then triggering the next cycle to search for the log message?

Reply:
@agrawalravi90 you are right in general. but here since we are not looking for job state in R but only X it is safe.

Reply:
My intent is to have some subjobs in 'X' state and some in 'Q' or 'R'. Also, to have the test not take a long time I reduced the time interval to 2 seconds. I'd like to retain as is.

Reply:
@arungrover if that is your intent then I suggest submit a job array requesting 100 subjobs or so.

Reply:
Ok, I can increase the number of subjobs here